File: app/src/main/java/com/pr0gramm/app/services/InviteService.java
package com.pr0gramm.app.services;

import android.util.Patterns;

import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.api.pr0gramm.response.AccountInfo;

import org.immutables.value.Value;

import java.util.List;
import java.util.regex.Matcher;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;

/**
 */
@Singleton
public class InviteService {
    private final Api api;

    @Inject
    public InviteService(Api api) {
        this.api = api;
    }

    public Observable<Void> send(String email) {
        Matcher matcher = Patterns.EMAIL_ADDRESS.matcher(email);
        if (!matcher.matches())
            return Observable.error(new InviteException("email"));

        return api.invite(null, email).flatMap(response -> {
            if (response.error() != null) {
                return Observable.error(new InviteException(response.error()));
            } else {
                return Observable.just(null);
            }
        });
    }

    public Observable<Invites> invites() {
        return api.accountInfo().map(info -> ImmutableInvites.builder()
                .inviteCount(info.account().invites())
                .invited(info.invited())
                .build());
    }

    public static class InviteException extends RuntimeException {
        private final String errorCode;

        public InviteException(String errorCode) {
            this.errorCode = errorCode;
        }

        public boolean noMoreInvites() {
            return ERROR_INVITES.equalsIgnoreCase(errorCode);
        }

        public boolean emailFormat() {
            return ERROR_EMAIL_FORMAT.equalsIgnoreCase(errorCode);
        }

        public boolean emailInUse() {
            return ERROR_IN_USE.equalsIgnoreCase(errorCode);
        }
    }

    @Value.Immutable
    public interface Invites {
        int inviteCount();

        List<AccountInfo.Invite> invited();
    }

    public static final String ERROR_EMAIL_FORMAT = "emailInvalid";
    public static final String ERROR_INVITES = "noInvites";
    public static final String ERROR_IN_USE = "emailInUse";
}

File: app/src/main/java/com/pr0gramm/app/services/InviteService.kt
package com.pr0gramm.app.services

import android.util.Patterns
import com.pr0gramm.app.api.pr0gramm.Api

/**
 */
class InviteService(private val api: Api) {
    /**
     * Send an invite to the given email address.
     */
    suspend fun send(email: String) {
        val matcher = Patterns.EMAIL_ADDRESS.matcher(email)
        if (!matcher.matches())
            throw InviteException("email")

        val response = api.inviteAsync(null, email).await()
        if (response.error != null) {
            throw InviteException(response.error)
        }
    }

    /**
     * Returns an observable producing the invites of the current user once.
     */
    suspend fun invites(): Invites {
        val info = api.accountInfoAsync().await()
        return Invites(info.account.invites, info.invited)
    }

    class InviteException(private val errorCode: String) : RuntimeException() {
        fun noMoreInvites(): Boolean {
            return "noInvites".equals(errorCode, ignoreCase = true)
        }

        fun emailFormat(): Boolean {
            return "emailInvalid".equals(errorCode, ignoreCase = true)
        }

        fun emailInUse(): Boolean {
            return "emailInUse".equals(errorCode, ignoreCase = true)
        }
    }

    data class Invites(val inviteCount: Int, val invited: List<Api.AccountInfo.Invite>)
}
